File: Python/DecoratorPattern/Coffee-Hut/coffee_hut.py
# This implementation is based on the example from Headfirst Design Patterns book
from beverage.beverage import Beverage
from beverage.condiments.mocha import Mocha
from beverage.condiments.soy import Soy
from beverage.condiments.whip import Whip
from beverage.dark_roast import DarkRoast
from beverage.espresso import Espresso
from beverage.house_blend import HouseBlend


def print_cost(beverage: Beverage):
    print(beverage.get_description() + " $" + str(beverage.cost()))


if __name__ == '__main__':
    # First Customer comes in to the Store and orders one plain House Blend Coffee
    order1: Beverage = HouseBlend()
    print_cost(order1)

    # Second customer orders an Espresso with double whip and soy milk
    order2: Beverage = Whip(Whip(Soy(Espresso())))
    print_cost(order2)

    # Third customer wants Dark Roast, with double mocha and whip
    order3: Beverage = Whip(Mocha(Mocha(DarkRoast())))
    print_cost(order3)

File: DecoratorPattern/Java/Coffee-Hut/src/CoffeeHut.java
import com.laxmena.beverage.Beverage;
import com.laxmena.beverage.DarkRoast;
import com.laxmena.beverage.Espresso;
import com.laxmena.beverage.HouseBlend;
import com.laxmena.beverage.condiments.Mocha;
import com.laxmena.beverage.condiments.Soy;
import com.laxmena.beverage.condiments.Whip;

// This Implementation is based on the example from HeadFirst Design Patterns book.
public class CoffeeHut {
    public static void main(String[] args) {
        // First Customer requests Espresso with no added condiments
        Beverage beverage = new Espresso();
        System.out.println(beverage.getDescription() + " $" + beverage.cost());

        // Second Customer requests double mocha whipped dark roast coffee
        Beverage beverage2 = new DarkRoast();
        beverage2 = new Mocha(beverage2); // Decorator is used to add additional functionality
        beverage2 = new Mocha(beverage2);
        beverage2 = new Whip(beverage2);
        System.out.println(beverage2.getDescription() + " $" + beverage2.cost());

        // Third customer requests House Blend with Soy, Mocha and Whip
        Beverage beverage3 = new HouseBlend();
        beverage3 = new Soy(beverage3);
        beverage3 = new Mocha(beverage3);
        beverage3 = new Whip(beverage3);
        System.out.println(beverage3.getDescription() + " $" + beverage3.cost());
    }
}
